layer/log_scanner.h

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef STADIA_OPEN_SOURCE_PERFORMANCE_LAYERS_LOG_SCANNER_H_
+#define STADIA_OPEN_SOURCE_PERFORMANCE_LAYERS_LOG_SCANNER_H_
+
+#include <cstdint>
+#include <fstream>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+
+namespace performancelayers {
+
+// Scans files looking for registered patterns. Assumes that log files will
+// be appended to, but not modified in the middle.
+class LogScanner {
+ public:
+  // Opens |filename| and returns a LogScanner object on success,
+  // or std::nullopt on failure.
+  static std::optional<LogScanner> from_filename(const std::string& filename);
+
+  // Reads the underlying log file until the end. Checks if any of the
+  // registered patterns appear within the lines read.
+  // Returns true if any register pattern was found for the first time.
+  bool consume_new_lines();
+
+  // Registers |pattern| as a watched pattern. No-op if pattern has
+  // already been registered.
+  // Patterns are matched using exact (full) maching only.
+  void register_watched_pattern(const std::string& pattern) {
+    pattern_to_line_num_.try_emplace(pattern, 0);
+  }
+
+  // Stops watching |pattern| and forgets the first match, if present.
+  void unregister_watched_pattern(const std::string& pattern);
+
+  // Returns the first line where |pattern| was seen, or 0 if not seen yet.
+  uint64_t get_first_occurrence_line_num(const std::string& pattern) const;
+
+  using PatternLineNumPair = std::pair<std::string, uint64_t>;
+  // Returns all seen patterns, sorted by the line of occurrence in asc. order.
+  std::vector<PatternLineNumPair> get_seen_patterns() const;
+
+  // Closes the underlying file. If not closed manually, the file will be
+  // automatically closed by the destructor.
+  void close() { file_.close(); }
+
+  // Returns true iff the underlying file is open.
+  bool is_open() const { return file_.is_open(); }
+
+ private:
+  LogScanner(std::ifstream file) : file_(std::move(file)) {}
+  std::ifstream file_;
+  uint64_t current_line_num_ = 0;
+  absl::flat_hash_map<std::string, uint64_t> pattern_to_line_num_;
```

Comment:
I don't think the layer needs to search for more than one pattern and does't need the line number (it only needs to know the pattern was found). Could the LogScanner API be simplified to something like the following?:
optional<LogScanner> scanner = LogScanner(file, pattern);
...
if (scanner.PatternFound()) {
  found_ = true;
  scanner.reset();
}

Reply:
I want to extend it to search for multiple patterns in a future PR. I found line numbers to be useful for debugging: they allow for jumping to the actual log file and looking for the pattern in the surrounding context.